rclcpp/include/rclcpp/create_generic_client.hpp

```diff
@@ -0,0 +1,56 @@
+// Copyright 2023 Sony Group Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__CREATE_GENERIC_CLIENT_HPP_
+#define RCLCPP__CREATE_GENERIC_CLIENT_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/generic_client.hpp"
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+#include "rclcpp/node_interfaces/node_graph_interface.hpp"
+#include "rclcpp/node_interfaces/node_services_interface.hpp"
+#include "rclcpp/qos.hpp"
+
+namespace rclcpp
+{
+/// Create a generic service client with a name of given type.
+/**
+ * \param[in] node_base NodeBaseInterface implementation of the node on which
+ *  to create the client.
+ * \param[in] node_graph NodeGraphInterface implementation of the node on which
+ *  to create the client.
+ * \param[in] node_services NodeServicesInterface implementation of the node on
+ *  which to create the client.
+ * \param[in] service_name The name on which the service is accessible.
+ * \param[in] service_type The name of service type, e.g. "test_msgs/srv/BasicTypes"
+ * \param[in] qos Quality of service profile for client.
+ * \param[in] group Callback group to handle the reply to service calls.
+ * \return Shared pointer to the created client.
+ */
+RCLCPP_PUBLIC
+rclcpp::GenericClient::SharedPtr
+create_generic_client(
+  std::shared_ptr<node_interfaces::NodeBaseInterface> node_base,
+  std::shared_ptr<node_interfaces::NodeGraphInterface> node_graph,
+  std::shared_ptr<node_interfaces::NodeServicesInterface> node_services,
+  const std::string & service_name,
+  const std::string & service_type,
+  const rclcpp::QoS & qos = rclcpp::ServicesQoS(),
+  rclcpp::CallbackGroup::SharedPtr group = nullptr);
```

Comment:
I think that in addition to this signature, we should have one that takes a single node argument and figures out the node interfaces for you, e.g.: https://github.com/ros2/rclcpp/blob/126d517193b6df8177680898e23906a85792eaf6/rclcpp/include/rclcpp/create_subscription.hpp#L168-L192

That is to say, I should be able to do something like:

```cpp
auto client = rclcpp::create_generic_client(my_weird_node_class_that_has_base_graph_and_service_interfaces, name, type, ...);
```

Reply:
@wjwwood 

After checking the Client's code, I found that it only offers one constructor as well. There is no constructor based on Node provided. In fact, users will not create a Client directly; it is always created through Node::create_client(). Therefore, the question arises whether a new constructor (with node argument) needs to be provided ?

Reply:
I see, that's likely to be the client APIs lagging behind the pub/sub. I do think we want a user-friendly `rclcpp::create_X(NodeT ...)` like free-function for each entity. You don't have to add a function like that for Client in this pr, but I do think it's a good thing to add for new code. If you don't add it for Client I can open a pr to do it separately.

Reply:
Get it.
I will add this new constructor with Node parameter. 
About new constructor for Client, it is better to use another PR to add.